src/fairseq2/checkpoint.py

```diff
@@ -0,0 +1,329 @@
+        """Save a checkpoint.
+
+        :param step_nr:
+            The training step number of the checkpoint.
+        :param state:
+            The checkpoint state.
+        """
+
+    @abstractmethod
+    def load_checkpoint(self, step_nr: int) -> Dict[str, Any]:
+        """Load the checkpoint of the specified training step.
+
+        :param step_nr:
+            The training step number.
+        """
+
+    @abstractmethod
+    def load_last_checkpoint(self) -> Tuple[int, Dict[str, Any]]:
+        """Load the last checkpoint.
+
+        :returns:
+            - The training step number of the checkpoint.
+            - The checkpoint state.
+        """
+
+    @abstractmethod
+    def get_last_step_nr(self) -> Optional[int]:
+        """Return the training step number of the last checkpoint."""
+
+    @abstractmethod
+    def has_last_checkpoint(self) -> bool:
+        """Return ``True`` if there is a checkpoint marked as last."""
+
+    @abstractmethod
+    def save_consolidated_fsdp_model(self, step_nr: int, model: Module) -> None:
+        """Save ``model`` with a consolidated ``state_dict``.
+
+        This method consolidates the state of the model from all ranks so that
+        it can be loaded on a single device without FSDP.
+        """
+
+    @abstractmethod
+    def load_consolidated_model(
+        self, step_nr: int, model: Module, device: Optional[Device] = None
+    ) -> None:
+        """Load the consolidated model of the specified training step.
+
+        :param step_nr:
+            The training step number.
+        :param model:
+            The model to load.
+        :param device:
+            The device on which to load the model if it is on a meta device.
+        """
+
+    @abstractmethod
+    def load_last_consolidated_model(
+        self, model: Module, device: Optional[Device] = None
+    ) -> int:
+        """Load the last consolidated model.
+
+        :param model:
+            The model to load.
+        :param device:
+            The device on which to load the model if it is on a meta device.
+
+        :returns:
+            The training step number of the consolidated model.
+        """
+
+
+@final
+class FileCheckpointManager(CheckpointManager):
+    """Saves and loads training checkpoints on a file system."""
+
+    checkpoint_dir: Path
+    gang: Gang
+
+    def __init__(self, checkpoint_dir: Path, gang: Gang) -> None:
+        """
+        :param checkpoint_dir:
+            The root directory under which to store the checkpoints.
+        """
+        self.checkpoint_dir = checkpoint_dir
+        self.gang = gang
+
+    @finaloverride
+    def save_checkpoint(self, step_nr: int, state: Dict[str, Any]) -> None:
+        checkpoint = {"step_nr": step_nr, "data": state}
+
+        step_dir = self.checkpoint_dir.joinpath(f"step_{step_nr}")
+
+        if self.gang.rank == 0:
+            try:
+                step_dir.mkdir(parents=True, exist_ok=True)
```

Comment:
When checkpointing FSDP models on non-nfs / local partitions, we may want to execute this step on all the nodes.

Reply:
What do you mean by "non-nfs/local" partitions?

Reply:
I was thinking of the case where we are checkpointing shards when using FSDP. In that case, all ranks will checkpoint their shard to the disk. We are creating the directory only in rank=0 process. Now if model was saved to a partition that is local to each node (i.e. physical machine), the checkpoint directory would only exist for the node with rank=0 process.

Reply:
Oh I see what you mean. In majority of cases, nodes share a common distributed file system (e.g NFS, fsx), but I think it is a fair ask to support local disk setup. Let me revise this implementation to address your feedback. Thanks!